core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceBillingAccountActor.scala

```diff
@@ -0,0 +1,275 @@
+   * sync the billing projects and all workspaces again immediately. If we sync one at a time
+   * we'll have a better chance of doing this once.
+   */
+  def updateBillingAccounts: IO[Unit] =
+    getABillingProjectChange.flatMap(_.traverse_ { billingAccountChange =>
+      for {
+        _ <- info("Updating Billing Account on Billing Project in Google",
+          "changeId" -> billingAccountChange.id,
+          "billingProject" -> billingAccountChange.billingProjectName.value,
+          "newBillingAccount" -> billingAccountChange.newBillingAccount.map(_.value)
+        )
+
+        billingProject <- loadBillingProject(billingAccountChange.billingProjectName)
+        billingProbeHasAccess <- billingProject.billingAccount.traverse(gcsDAO.testDMBillingAccountAccess(_).io)
+        billingProjectSyncAttempt <- syncBillingProjectWithGoogle(billingProject).attempt
+        syncOutcome = Outcome.fromEither(billingProjectSyncAttempt)
+        _ <- recordBillingProjectSyncOutcome(billingAccountChange, billingProbeHasAccess, syncOutcome)
+        _ <- updateWorkspacesInProject(billingAccountChange, billingProject, syncOutcome)
+      } yield ()
+    })
+
+
+  def getABillingProjectChange: IO[Option[BillingAccountChange]] =
+    dataSource
+      .inTransaction(_ => BillingAccountChanges.latestChanges.unsynced.take(1).result).io
+      .map(_.headOption)
+
+
+  private def loadBillingProject(projectName: RawlsBillingProjectName): IO[RawlsBillingProject] =
+    dataSource
+      .inTransaction( _ => rawlsBillingProjectQuery.load(projectName))
+      .io
+      .map(_.getOrElse(throw new RawlsException(s"No such billing account $projectName")))
+
+
+  private def syncBillingProjectWithGoogle(project: RawlsBillingProject): IO[Unit] =
+    for {
+      // only v1 billing projects are backed by google projects
+      isV1BillingProject <- gcsDAO.rawlsCreatedGoogleProjectExists(project.googleProjectId).io
+      _ <- Applicative[IO].whenA(isV1BillingProject)(
+        updateBillingAccountOnGoogle(project.googleProjectId, project.billingAccount)
+      )
+    } yield ()
+
+
+  private def recordBillingProjectSyncOutcome(change: BillingAccountChange,
+                                              billingProbeCanAccessBillingAccount: Option[Boolean],
+                                              outcome: Outcome): IO[Unit] =
+    for {
+      syncTime <- IO(Instant.now)
+
+      baseInfo = Seq(
+        "changeId" -> change.id,
+        "billingProject" -> change.billingProjectName.value,
+        "newBillingAccount" -> change.newBillingAccount.map(_.value)
+      )
+
+      (_, message) = Outcome.toTuple(outcome)
+      _ <- if (outcome.isSuccess)
+        info("Successfully updated Billing Account on Billing Project in Google", baseInfo :_*) else
+        warn("Failed to update Billing Account on Billing Project in Google", ("details" -> message) +: baseInfo :_*)
+
```

Comment:
When I first read this message, I thought this was recording the result of syncing all google projects (GP) in a billing project (BP) (including any v2 workspaces' GPs). This message is logged after syncing a v1 BP, but _before_ syncing v2 workspaces. 

If the intention was to print this message after all GPs (in a BP) are synced, I think we need to move this function so it happens after the workspace sync. Else maybe we should make the log say explicitly that this is for v1 BPs.

Q: If a v2 workspace GP sync fails, shouldn't that affect the Outcome of the BillingAccountChange?

Reply:
Yep. Good point. I incorporaed your suggestion.